reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java

```diff
@@ -226,6 +226,95 @@ void onComplete() {
 			parent.remove(this);
 			actual.onComplete();
 		}
+
+		private void drain() {
+			if (WIP.getAndIncrement(this) == 0) {
+				while (WIP.get(this) > 0) {
```

Comment:
you need to perform that check _atomically_ in the `decrementAndGet` at the end of the loop, rather than here, otherwise it defeats the purpose:
 - thread A wins the guard and goes into the loop
 - thread B looses and exits
 - thread A does a first loop, decrements
 - thread A does a second loop, corresponding to thread B's event
 - **thread A decrements to 0**
 - **thread C enters `drain` and wins the guard (swapping from 0 to 1)**
 - **thread A reads `WIP` and sees `1`**
 - **:bug: both A and C enter the loop and clash**